### PR TITLE
[Utilities] Added ManagementGroup-Deployment removal script

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

- Created a utility to remove any non-running deployment from a given Management-Group scope
- Uses batch processing, which let's you remove the deployments within a few seconds.
- This utility can later be used in issue #1320 , but until we automate it, it's useful to have around for manual execution

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+function Clear-ManagementGroupDeployment {
+
+
+    [CmdletBinding(SupportsShouldProcess)]
+    param (
+        [Parameter(Mandatory = $true)]
+        [string] $ManagementGroupId,
+
+        [Parameter(Mandatory = $false)]
+        [string] $DeploymentStatusToExclude = 'running'
```
Comment (eriqua) [inline utilities/tools/Clear-ManagementGroupDeployment.ps1:29]: Is there a use case for which we want to remove running deployments? If not, I'd avoid this option, removing only completed deployments, maybe updating the function and .ps file names to reflect this (e.g. Clear-ManagementGroupCompletedDeployment)

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    )
```
Comment (eriqua) [inline utilities/tools/Clear-ManagementGroupDeployment.ps1:30]: I'd suggest to add a boolean parameter to allow skipping the removal of failed deployments, useful for troubleshooting. E.g.
``` 
        [Parameter(Mandatory = $false)]
        [bool] $SkipFailedDeployments = $true
```

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    . (Join-Path $PSScriptRoot 'helper' 'Split-Array.ps1')
+
+    $getInputObject = @{
+        Method  = 'GET'
+        Uri     = "https://management.azure.com/providers/Microsoft.Management/managementGroups/$ManagementGroupId/providers/Microsoft.Resources/deployments/?api-version=2021-04-01"
+        Headers = @{
+            Authorization = 'Bearer {0}' -f (Get-AzAccessToken).Token
+        }
+    }
+    $response = Invoke-RestMethod @getInputObject
```
Comment (eriqua) [inline utilities/tools/Clear-ManagementGroupDeployment.ps1:42]: Why not using `Invoke-AzRestMethod`? Cleaner code avoiding Headers and consistent with what we use for removal.

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        }
+        if ($PSCmdlet.ShouldProcess(('Removal of [{0}] deployments' -f $requests.Count), 'Request')) {
+            $response = Invoke-RestMethod @removeInputObject
+
+            $failedRemovals += ($response.responses | Where-Object { $_.httpStatusCode -notlike '20*' }  ).Count
+            $successfulRemovals += ($response.responses | Where-Object { $_.httpStatusCode -like '20*' }  ).Count
+        }
+    }
+
+    Write-Verbose 'Outcome' -Verbose
```
Comment (eriqua) [inline utilities/tools/Clear-ManagementGroupDeployment.ps1:92]: Wondering if useful to add an option to list removed deployment names + deployment status, maybe disabled by default. Not necessarily in this PR.

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    Write-Verbose ('Triggering the removal of [{0}] deployments of management group [{1}]' -f $relevantDeployments.Count, $ManagementGroupId)
+
+    $failedRemovals = 0
+    $successfulRemovals = 0
+    foreach ($deployments in $relevantDeploymentChunks) {
+
+        $requests = $deployments | ForEach-Object {
+            @{ httpMethod            = 'DELETE'
+                name                 = (New-Guid).Guid # $_.properties.correlationId # "73c46fa8-dffe-4b9e-9360-cd0dc3c7b4d8",
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1]: removed

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 3 above)
Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:42]: I tried - and failed. It returns error messages above subscriptions not being set - which is ironic as I want to execute actions on an above subscription-level. Troubleshooting didn't return and results so I went with the general and working `Invoke-RestMethod`.

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 4 above)
Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:92]: It is a good question. Though the more I think about it, the more I wonder what you could do with that kind of information. It's not like you could recover the deployments anyways after they're deleted ^^

Comment (eriqua) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:92]: You sure won't but could find useful to know if past deployment you removed failed or not. Probably best would be to also return the deployment error in failure cases. BTW something we could discuss as an improvement of this already useful version.

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 3 above)
Comment (eriqua) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:42]: disappointing but fine :)

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+.DESCRIPTION
+Bulk delete all deployments on the given management group scope
+
+.PARAMETER ManagementGroupId
+Mandatory. The Resource ID of the Management Group to remove the deployments for.
+
+.PARAMETER DeploymentStatusToExclude
+Optional. The status to exlude from removals. Can be multiple. By default, we exclude any deployment that is in state 'running'.
+
+.EXAMPLE
```
Comment (eriqua) [inline utilities/tools/Clear-ManagementGroupDeployment.ps1:15]: I'd suggest to add an example for `DeploymentStatusToExclude `usage

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 2 above)
Comment (eriqua) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:30]: Now that it is an array, sure. I'm anyway still in doubt if we shouldn't keep failed deployments by default. I'd go for it but it's also true we could eventually reach the same 800 limit with failed deployments only. What do you think?

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 [hunk 7]
```diff
@@ -0,0 +1,96 @@
+﻿
```
Comment (eriqua) [inline utilities/tools/Clear-ManagementGroupDeployment.ps1:1]: We should probably add a wiki page for this as done for other utilities.

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 2 above)
Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:30]: Folks will eventually run of ouf deployments and need to remove failed ones too. However, I'm fine with keeping them by default until this limit is actually reached. I guess the pipeline we want as part of CARML could have a flag that deletes them too if a user reaaaally wants to.

Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:30]: done

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 4 above)
Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:92]: Agreed

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 6 above)
Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:15]: done

File: utilities/tools/Clear-ManagementGroupDeployment.ps1 (hunk 7 above)
Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:1]: Was thinking about this - but I couldn't find a place where to put it. It was easy with the original Wiki structure as we had the 'utilities' parent folder, but now?
I eventually ditched the idea as I figured that we'll eventually invoke it via a scheduled pipeline in the other issue and it's not like we documented the other pipeline scripts ^^. Thoughts?

Comment (eriqua) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:1]: I think it naturally belongs to the CI environment, deployment validation in particular. It makes sense to be used only after deploying modules at mg scope.

Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:1]: Alternatively - Known issues...

Comment (MrMCake) [reply utilities/tools/Clear-ManagementGroupDeployment.ps1:1]: Added

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**